noir-projects/noir-contracts/contracts/nft_contract/src/main.nr

```diff
@@ -0,0 +1,291 @@
+
+    #[aztec(public)]
+    #[aztec(view)]
+    fn public_get_symbol() -> pub FieldCompressedString {
+        storage.symbol.read_public()
+    }
+
+    #[aztec(private)]
+    #[aztec(view)]
+    fn private_get_symbol() -> pub FieldCompressedString {
+        storage.symbol.read_private()
+    }
+
+    #[aztec(public)]
+    #[aztec(view)]
+    fn get_admin() -> Field {
+        storage.admin.read().to_field()
+    }
+
+    #[aztec(public)]
+    #[aztec(view)]
+    fn is_minter(minter: AztecAddress) -> bool {
+        storage.minters.at(minter).read()
+    }
+
+    #[aztec(public)]
+    fn transfer_in_public(from: AztecAddress, to: AztecAddress, token_id: Field, nonce: Field) {
+        if (!from.eq(context.msg_sender())) {
+            assert_current_call_valid_authwit_public(&mut context, from);
+        } else {
+            assert(nonce == 0, "invalid nonce");
+        }
+
+        let public_owners_storage = storage.public_owners.at(token_id);
+        assert(public_owners_storage.read().eq(from), "invalid owner");
+
+        public_owners_storage.write(to);
+    }
+
+    /// Prepares a transfer from public balance of `from` to a private balance of `to`. The transfer then needs to be
+    /// finalized by calling `finalize_transfer_to_private`. `transient_storage_slot_randomness` is passed
+    /// as an argument so that we can derive `transfer_preparer_storage_slot_commitment` off-chain and then pass it
+    /// as an argument to the followup call to `finalize_transfer_to_private`.
+    // TODO(#8238): Remove the `note_randomness` argument below once we have partial notes delivery (then we can just
+    // fetch the randomness from oracle).
+    #[aztec(private)]
+    fn prepare_transfer_to_private(
+        from: AztecAddress,
+        to: AztecAddress,
+        note_randomness: Field,
+        transient_storage_slot_randomness: Field
+    ) {
+        // We create a partial NFT note hiding point with unpopulated/zero token id for 'to'
+        let to_npk_m_hash = get_current_public_keys(&mut context, to).npk_m.hash();
+        let to_note_slot = storage.private_nfts.at(to).storage_slot;
+        let hiding_point = NFTNoteHidingPoint::new(to_npk_m_hash, to_note_slot, note_randomness);
+
+        // We make the msg_sender/transfer_preparer part of the slot preimage to ensure he cannot interfere with
+        // non-sender's slots
+        let transfer_preparer_storage_slot_commitment: Field = pedersen_hash(
+            [context.msg_sender().to_field(), transient_storage_slot_randomness],
+            TRANSIENT_STORAGE_SLOT_PEDERSEN_INDEX
+        );
+        // Then we hash the transfer preparer storage slot commitment with `from` and use that as the final slot
+        // --> by hashing it with a `from` we ensure that `from` cannot interfere with slots not assigned to him.
+        let slot: Field = pedersen_hash(
+            [from.to_field(), transfer_preparer_storage_slot_commitment],
+            TRANSIENT_STORAGE_SLOT_PEDERSEN_INDEX
```

Comment:
So I am mixing `from` into the slot here to essentially obviate the need for public authwits in `finalize_transfer_to_private` as it ensures that only `from` can access the prepared transfer in `finalize_transfer_to_private`. If we instead relied on public authwits this would not be necessary but that would require a call to `AuthRegistry`.